client/chatting.ts

```diff
@@ -0,0 +1,401 @@
+	const codebook = getCodebook();
+	codebook.messageKeypair.set(messageId, { publicKey, privateKey });
+	localStorage.setItem(getUser() + '-keys', JSON.stringify(codebook, replacer));
+}
+
+// Retrieves a keypair from the codebook
+function loadKeypair (messageId: string): Keypair | undefined {
+	const codebook = getCodebook();
+	return codebook.messageKeypair.get(messageId);
+}
+
+// Sends the ANNOUNCE chat message component to server
+function sendAnnounce (recipientId: string, messageContent: string): void {
+	const payload = {
+		messageId: null,
+		userId: recipientId,
+		stage: 'ANNOUNCE',
+		content: 'a|b' // Used to specify common values for Diffie-Hellman, now useless due to ECDH. Easier to leave it with a dummy value
+	};
+	fetchWithCSRF('/chat', {
+		method: 'POST',
+		body: JSON.stringify(payload)
+	}).then(async res => {
+		if (res.status !== 200) {
+			console.error(`ANNOUNCE message returned ${res.status}`);
+		} else {
+			if (res.body !== null) {
+				// Success, we can queue the message, send out public key and wait for the other side to send theirs
+				const payload = await res.text();
+				queueMessage(payload, messageContent);
+				saveMessage(payload, messageContent);
+				sendKeyComponent(payload);
+			}
+		}
+	}).catch(e => {
+		console.error(`ANNOUNCE message to ${recipientId} failed: ${e as string}`);
+	});
+}
+
+// Sends the client's public key to the server
+function sendKeyComponent (messageId: string): void {
+	const user = getUser();
+	const dh = createDiffieHellman('secp256k1'); // Create a ECDH algorithm instance, the specific curve does not really matter, secp256k1 is good enough.
+	const preexistingKey = loadKeypair(messageId); // Check whether a keypair has already been generated (otherwise we might accidentally discard the working key)
+	if (preexistingKey !== undefined) {
+		awaitKey(messageId); // Mark down the message as awaiting key, since it seems to not have been marked as one already
+		return;
+	}
+	dh.generateKeys(); // Generate a keypair
+	const keycomp = dh.getPublicKey('hex') as string; // Prepare the public key as a hex string.
+	saveKeypair(messageId, keycomp, dh.getPrivateKey('hex') as string); // Save both keys for later usage
+	const payload = {
+		messageId,
+		userId: null,
+		stage: 'KEY',
+		content: user + '|' + keycomp // Sender ID | Public Key
+	};
+	fetchWithCSRF('/chat', {
+		method: 'POST',
+		body: JSON.stringify(payload)
+	}).then(async res => {
+		if (res.status !== 200) {
+			console.error(`KEY message returned ${res.status}`);
+		} else {
+			// If successful, await key from other side
+			awaitKey(messageId);
+		}
+	}).catch(e => {
+		console.error(`KEY message to ${messageId} failed: ${e as string}`);
+	});
+}
+
+// Retrieves the other user's public key from the server
+async function fetchKey (messageId: string): Promise<void> {
+	const response = await fetch(`/chat/message/${messageId}?stage=KEY`);
+	if (response.status !== 200) return; // Key is not ready, we need to wait
```

Comment:
You might notice that I do not actually queue an another request. 
If it's unsuccessful, it just returns the function and it will not be called again until a state is changed (i.e. the page is refreshed or the socket gets a messageKey event). Hence, it doesn't actually spam the server with pointless requests.

Reply:
eh still, checking with appropriate status numbers is good practice :)

Reply:
Not really, the function should no continue at any other status code either. Any non-200 code implies that either what the function is asking for does not (yet) exist, or something has gone awfully wrong. In both scenarios, continuing is impossible.